library(data.table)
library(tidyverse)
library(DESeq2)
library(ChIPQC)
library(rtracklayer)
library(DT)
library(dplyr)
library(tidyr)
library(GenomicRanges)
library(DiffBind)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(ATACseqQC)
library(data.table)
library(dplyr)
library(tximeta)
library(GenomicAlignments)
library(ggpubr)
library(ChIPpeakAnno)

args.n <- commandArgs(TRUE)


#shifting aligned read-Tagmentation of Tn5 transposases produce 5â€™ overhang of 9 base long, the coordinates of reads mapping to the positive and negative strands need to be shifted by + 4 and - 5, respectively
possibleTag <- combn(LETTERS, 2)
possibleTag <- c(paste0(possibleTag[1, ], possibleTag[2, ]),
                 paste0(possibleTag[2, ], possibleTag[1, ]))
bamFile <- list.files("../align",pattern = "nodups_filtered_sorted.bam$", full.names = TRUE)
bamFileLabels <- gsub("_18_nodups_filtered_sorted.bam", "", basename(bamFile))

## run program for nucleosome-free region and region occupied mononucleosome dinucleosome, and trinucleosomes
seqlev <- sprintf("chr%s",c(seq(1:22),"X","Y"))
which <- as(seqinfo(Hsapiens)[seqlev], "GRanges")
txs <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
txs <- txs[seqnames(txs) %in% seqlev]
genome <- Hsapiens

i <- as.integer(args.n[1])
bamTop100 <- scanBam(BamFile(bamFile[i], yieldSize=1e+6),
                     param = ScanBamParam(tag = possibleTag))[[1]]$tag
tags <- names(bamTop100)[lengths(bamTop100)>0]

## files will be output into outPath
outPath <- paste0("splitBam/",bamFileLabels[i])
if (dir.exists(outPath))
{
  unlink(outPath, recursive = TRUE, force = TRUE)
}
dir.create(file.path("splitBam",bamFileLabels[i]), recursive = TRUE)
## shift the coordinates of 5'ends of alignments in the bam file

gal <- readBamFile(bamFile[i], tag=tags, which=which,
                   asMates=TRUE, bigFile=FALSE)
shiftedBamFile <- file.path(outPath, paste0(bamFileLabels[i],"_","shifted.bam"))
gal1 <- shiftGAlignmentsList(gal, outbam=shiftedBamFile)



## split the reads into bins for reads derived from nucleosome-free regions, mononucleosome, dinucleosome and trinucleosome. And save the binned alignments into bam files.
objs <- splitGAlignmentsByCut(gal1, txs=txs, genome=genome, outPath = outPath)
## list the files generated by splitGAlignmentsByCut.
dir(outPath)
lst <- list.files(paste0("splitBam/",bamFileLabels[i]), pattern = ".bam$")
sapply(lst,function(x) file.rename(from =paste0(paste0("splitBam/",bamFileLabels[i],"/"),x), to=paste0(paste0("splitBam/",bamFileLabels[i],"/"),bamFileLabels[i],"_",x)))
lst <- list.files(paste0("splitBam/",bamFileLabels[i]), pattern = ".bai$")
sapply(lst,function(x) file.rename(from =paste0(paste0("splitBam/",bamFileLabels[i],"/"),x), to=paste0(paste0("splitBam/",bamFileLabels[i],"/"),bamFileLabels[i],"_",x)))




bamFiles <- list.files(outPath,pattern = "bam$", full.names = TRUE)

TSS <- promoters(txs, upstream=0, downstream=1)
TSS <- unique(TSS)
## estimate the library size for normalization
(librarySize <- estLibSize(bamFiles))

#atac_count <- fread("../peaks/CQTL_18_peakCounts.tsv",fill=TRUE, drop= c("V10"))
#chr_nm <- sprintf("chr%s",c(seq(1:22),"X","Y"))
#atac_ct_subs <- atac_count[atac_count$chr %in% chr_nm,]

#select(atac_ct_subs,contains("CTL"))

